File: backend/utils/distance.js
// Haversine distance in km between two lat/lng points.
// Returns null if either point is missing coordinates (MVP: distance is "unknown" then,
// UI should fall back to showing locality name only).
function haversineKm(pointA, pointB) {
  if (!pointA || !pointB || pointA.lat == null || pointA.lng == null || pointB.lat == null || pointB.lng == null) {
    return null;
  }
  const R = 6371; // Earth radius km
  const dLat = toRad(pointB.lat - pointA.lat);
  const dLng = toRad(pointB.lng - pointA.lng);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(pointA.lat)) * Math.cos(toRad(pointB.lat)) * Math.sin(dLng / 2) * Math.sin(dLng / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return Math.round(R * c * 10) / 10; // 1 decimal place
}

function toRad(deg) {
  return (deg * Math.PI) / 180;
}

module.exports = { haversineKm };

File: backend/app/utils/distance.py
import math

def haversine_km(point_a: dict | None, point_b: dict | None) -> float | None:
    """
    Haversine distance in km between two lat/lng points.
    Returns None if either point is missing coordinates.
    """
    if not point_a or not point_b:
        return None
    
    lat1 = point_a.get("lat")
    lng1 = point_a.get("lng")
    lat2 = point_b.get("lat")
    lng2 = point_b.get("lng")

    if lat1 is None or lng1 is None or lat2 is None or lng2 is None:
        return None

    try:
        lat1, lng1, lat2, lng2 = float(lat1), float(lng1), float(lat2), float(lng2)
    except (ValueError, TypeError):
        return None

    r = 6371.0  # Earth radius in km
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)
    a = (
        math.sin(d_lat / 2.0) ** 2
        + math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) * math.sin(d_lng / 2.0) ** 2
    )
    c = 2.0 * math.atan2(math.sqrt(a), math.sqrt(1.0 - a))
    return round(r * c, 1)
